libuavcan/include/uavcan/util/math.hpp

```diff
@@ -0,0 +1,227 @@
+/*
+ * Helpers for doing mathy things that aren't provided by the standard library.
+ *
+ * Copyright (C) 2014 Pavel Kirienko <pavel.kirienko@gmail.com>
+ * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ */
+/** @file */
+
+#ifndef UAVCAN_UTIL_MATH_HPP_INCLUDED
+#define UAVCAN_UTIL_MATH_HPP_INCLUDED
+
+#include "uavcan/uavcan.hpp"
+#include <cstdlib>
+#include <cmath>
+#include "uavcan/introspection.hpp"
+
+namespace uavcan
+{
+/**
+ * @namespace util
+ * Contains helper types and functions to make integrating libuavcan easier.
+ */
+namespace util
+{
+// +--------------------------------------------------------------------------+
+// | GENERIC SATURING MATH
+// +--------------------------------------------------------------------------+
+
+/*
+ * Saturated addition where the RHS is assumed to be either a positive value or, if negative, a signal to
+ * add "max + 1" to the lhs.
+ *
+ * This is part of the generic, promotionless, saturation math provided by this header. Ideally this would
+ * be omitted for most platforms where saturation operations were available.
+ */
+template <
+    class SIGNED_TYPE,
+    typename std::enable_if<std::is_integral<SIGNED_TYPE>::value && std::is_signed<SIGNED_TYPE>::value, int>::type = 0>
+SIGNED_TYPE _saturating_add_d(SIGNED_TYPE left, SIGNED_TYPE right)
```

Comment:
I hope you appreciate how I am purposefully saying nothing about bad naming here. I am writing this to say that this function can probably be made constexpr. I am saying "probably" because constexpr functions are somewhat limited in C++11, and I'm not that much familiar with those limitations since I'm mostly dealing with C++17 here these days.

Reply:
I can make some of these constexpr by rewriting them using a ternary operator. For the ones that cannot be rewritten like this should I add, perhaps:
```
future_constexpr 
```
which will be constexpr when compiling using c++14 or greater? I'm not sure how much effort we want to put into forward optimizations given that we'll need to add additional builds for each dialect we want to support.

Reply:
Let us not trade correctness and robustness for resource utilization.